crates/wasmtime/src/runtime/vm/memory/mmap.rs

```diff
@@ -178,23 +190,20 @@ impl RuntimeLinearMemory for MmapMemory {
             // or "dynamic" heaps which have some space reserved after the
             // initial allocation to grow into before the heap is moved in
             // memory.
-            assert!(new_size > self.len);
+            assert!(new_size <= current_capacity.byte_count());
             assert!(self.maximum.map_or(true, |max| new_size <= max));
-            assert!(new_size <= self.mmap.len() - self.offset_guard_size - self.pre_guard_size);
-
-            let new_accessible = round_usize_up_to_host_pages(new_size)?;
-            assert!(
-                new_accessible <= self.mmap.len() - self.offset_guard_size - self.pre_guard_size,
-            );
 
             // If the Wasm memory's page size is smaller than the host's page
             // size, then we might not need to actually change permissions,
             // since we are forced to round our accessible range up to the
             // host's page size.
-            if new_accessible > self.accessible() {
+            if let Ok(difference) = new_accessible.checked_sub(self.accessible()) {
```

Comment:
I think this may be the source of the bug, albeit it's a bit indirect. Previously the `>` condition is now morally becoming `>=` since `new_accessible == self.accessible()` didn't run this block before but it's now running this block with `difference == 0`. I think then it may be the case that mprotect for 0 bytes is succeeding on unix but failing on Windows.

On Linux I see `mprotect(0x740d36001000, 0, PROT_READ|PROT_WRITE) = 0` in strace with this patch when running the failing spec test and while I haven't executed this on Windows this is what I would suspect is going on.

Another option is to update the `Mmap` type to skip `make_accessible` when the byte size is zero which I think would also be reasonable.

Reply:
Oh that's terrifying -- I'm going to fix this (probably by making `make_accessible` ignore zero-sized regions) and go through this code again to make sure I didn't miss any other sign changes. Thanks for diagnosing this!

Reply:
Made `make_accessible` a no-op on zero-length allocations, as well as the other mprotect calls (`make_executable` and `make_readonly`). They also had this as a lurking bug because they had assertions that `range.start <= range.end`, not `range.start < range.end`.

I've also added tests for zero-length calls, and re-reviewed everything -- didn't see any other changes.